File: reference-harness/src/reference_harness/providers/mariadb.py
"""Testcontainers-backed MariaDB provider (dialect = "mariadb").

The MariaDB concrete dialect behind the m-db-port database-provider seam,
proving "equivalent SQL per database, optimized per dialect" beyond Postgres.
MariaDB is fully open-source and exercises exactly the seam points Postgres does
not:

* **Read-lock divergence.** MariaDB has no ``for share``; the shared row lock is
  ``lock in share mode`` (MDEV-17514). The golden SQL carries that form per
  dialect; the m-sql normalizer renders it through the seam.
* **No native timestamp infinity.** MariaDB's ``DATETIME`` has no ``'infinity'``,
  so the open temporal upper bound (m-core/m-dialect) maps to a documented **max-sentinel**
  — ``9999-12-31 23:59:59.999999`` (the largest ``DATETIME(6)``). This provider
  translates the declared temporal-infinity marker to that sentinel on the way in
  (fixture loads, binds) and back to ``"infinity"`` on the way out (reads), so a
  fixture authored once against native-infinity Postgres compares identically
  here — the only place the difference is allowed to surface.

It is booted via testcontainers-python's MySQL-compatible container pointed at a
``mariadb`` image, driven by the ``pymysql`` client.
"""

from __future__ import annotations

import datetime as _dt
import json
import time
import uuid
from collections.abc import Iterator, Sequence
from contextlib import contextmanager
from types import TracebackType
from typing import TYPE_CHECKING, Any

import pymysql
from pymysql.constants import FIELD_TYPE
from testcontainers.community.mysql import MySqlContainer

from .. import errors
from .._declared_contributor import TEMPORAL_INFINITY
from ..ddl_builder import quote_identifier
from ..document_codec import is_document
from . import register
from ._binds import adapt_document_scalar_binds

if TYPE_CHECKING:
    from . import Node

# The provisioning seam pins a MariaDB long-term-support series (m-case-format/DQ15).
# This pin must stay at or above 11.4.2, the first release carrying
# `innodb_snapshot_isolation`, without which Repeatable Read cannot forbid the lost
# update; the compatibility suite asserts the booted server against that floor.
MARIADB_IMAGE = "mariadb:12.3"

# MariaDB keys on the vendor errno; a code absent here is `errors.UNKNOWN`, so an
# unclassified error fails an assertion loudly rather than passing silently.
_CODES: dict[int, str] = {
    1062: errors.UNIQUE_VIOLATION,  # ER_DUP_ENTRY
    1213: errors.DEADLOCK,  # ER_LOCK_DEADLOCK
    1205: errors.LOCK_WAIT_TIMEOUT,  # ER_LOCK_WAIT_TIMEOUT
    # ER_CHECKREAD: a snapshot-isolation write/write conflict at Repeatable Read,
    # which the server itself rolls back like a deadlock.
    1020: errors.DEADLOCK,
}

# MariaDB has no native timestamp infinity; the open upper bound (m-core/m-dialect) is the
# largest representable DATETIME(6). This is the documented max-sentinel the seam
# substitutes for the suite's `infinity` literal.
_INFINITY_SENTINEL = _dt.datetime(9999, 12, 31, 23, 59, 59, 999999)
_INFINITY_LITERAL = "infinity"


def _to_db_bind(value: Any) -> Any:
    """Adapt a fixture / bind value for MariaDB binding.

    * the declared Timestamp infinity marker -> the max-sentinel ``DATETIME``
      (m-dialect);
    * a managed aware ``datetime`` -> naive UTC for MariaDB ``DATETIME``;
    * a portable document (``document_codec.is_document``) -> JSON text, because
      MariaDB's ``JSON`` is a text alias; serialization is this seam's job and sits
      below the bind, which is why a golden authors the document itself
      (m-case-format);
    * every other scalar passes through unchanged.
    """
    if value is TEMPORAL_INFINITY:
        return _INFINITY_SENTINEL
    if isinstance(value, _dt.datetime) and value.tzinfo is not None:
        return value.astimezone(_dt.UTC).replace(tzinfo=None)
    if is_document(value):
        return json.dumps(value)
    if isinstance(value, uuid.UUID):
        return str(value)
    return value


def _statement_binds(sql: str, binds: Sequence[Any]) -> tuple[Any, ...]:
    physical = adapt_document_scalar_binds(sql, binds, "mariadb")
    return tuple(_to_db_bind(value) for value in physical)


def _is_boolean_field(type_code: Any, column_length: Any) -> bool:
    """True for a MariaDB ``tinyint(1)`` result column — the ``boolean`` mapping.

    ``boolean`` is the ONLY neutral type that maps to ``tinyint(1)`` (ddl_builder),
    which pymysql reports as ``FIELD_TYPE.TINY`` with a display length of ``1`` and
    reads back as int ``0`` / ``1``. Every other integer column is a wider type code
    (``int32`` -> ``int`` / ``FIELD_TYPE.LONG``, ``int64`` -> ``bigint`` /
    ``FIELD_TYPE.LONGLONG``), so this never coerces a non-boolean integer.
    """
    return type_code == FIELD_TYPE.TINY and column_length == 1


def _from_db_value(value: Any, *, is_boolean: bool = False) -> Any:
    """Adapt a MariaDB-read value back to the suite's canonical form.

    * a ``tinyint(1)`` (``boolean``) column -> a real ``bool``: MariaDB has no
      native boolean, so pymysql reads the column back as int ``0`` / ``1``. The
      row comparator keeps ``bool`` OUT of numeric space (m-case-format layer 2:
      ``true`` is never ``1``), so an int would never match the fixture's boolean;
      the caller passes ``is_boolean`` from the column's ``tinyint(1)`` field
      metadata (:func:`_is_boolean_field`) so the value compares correctly. A NULL
      boolean column stays ``None``;
    * the max-sentinel ``DATETIME`` -> the literal ``"infinity"`` (m-dialect), so a
      current-row open bound compares to the fixture's ``infinity``;
    * a finite ``datetime`` -> a stable ISO-8601 UTC string with ``+00:00`` (the
      same shape the Postgres provider yields), so instant columns compare across
      dialects;
    * a bare ``date`` -> passed through unchanged as a ``datetime.date``,
      symmetric with the Postgres provider (which registers no ``date`` loader)
      and matching how a YAML-authored date scalar parses, so a ``date`` column
      compares equal across dialects; a ``bytes`` JSON payload is left to the
      caller (we never read JSON columns back for comparison in this phase).
    """
    if is_boolean and isinstance(value, int) and not isinstance(value, bool):
        return bool(value)
    if isinstance(value, _dt.datetime):
        if value == _INFINITY_SENTINEL:
            return _INFINITY_LITERAL
        return value.replace(tzinfo=_dt.UTC).isoformat()
    if isinstance(value, _dt.timedelta):
        # pymysql reads a `TIME` column as a `timedelta`; render it as a stable
        # `HH:MM:SS` string so a `time` column compares to a plain YAML string
        # (the Postgres provider already reads `time` as text). Sub-day values —
        # the only ones the suite carries — format cleanly via str().
        return str(value)
    return value


def _decode_rows(
    description: Sequence[Any],
    rows: Sequence[Sequence[Any]],
) -> list[dict[str, Any]]:
    """Decode a MariaDB cursor's fetched rows to the suite's canonical dicts.

    Reads each result column's ``tinyint(1)`` (``boolean``) flag from the cursor
    ``description`` once, then adapts every value through :func:`_from_db_value`.
    Shared verbatim by :meth:`MariaDbProvider.query` and :meth:`_MariaTxSession.query`
    so a future canonicalization change updates one decoding path.
    """
    columns = [(desc[0], _is_boolean_field(desc[1], desc[3])) for desc in description]
    return [
        {
            name: _from_db_value(value, is_boolean=is_boolean)
            for (name, is_boolean), value in zip(columns, row, strict=True)
        }
        for row in rows
    ]


class MariaDbProvider:
    """A clean, migrated, isolated MariaDB database for one suite run."""

    dialect = "mariadb"

    def __init__(
        self,
        connection: pymysql.connections.Connection,
        dbname: str,
        connect_params: dict[str, Any] | None = None,
    ) -> None:
        self._conn = connection
        self._dbname = dbname
        # Connection parameters for opening a second, independent connection to the
        # SAME database. A peer is created
        # by reconnecting with these and selecting the working database.
        self._connect_params = connect_params or {}

    # --- DatabaseProvider seam ---------------------------------------------

    def reset(self) -> None:
        """Drop and recreate the working schema — a clean, empty database."""
        with self._conn.cursor() as cur:
            cur.execute(f"drop database if exists {self._dbname}")
            cur.execute(f"create database {self._dbname}")
            cur.execute(f"use {self._dbname}")
        self._conn.commit()

    def apply_ddl(self, statements: Sequence[str]) -> None:
        with self._conn.cursor() as cur:
            for statement in statements:
                cur.execute(statement)
        self._conn.commit()

    def load(
        self,
        table: str,
        columns: Sequence[str],
        rows: Sequence[Sequence[Any]],
    ) -> None:
        if not rows:
            return
        col_list = ", ".join(quote_identifier(column, self.dialect) for column in columns)
        placeholders = ", ".join(["%s"] * len(columns))
        target = quote_identifier(table, self.dialect)
        sql = f"insert into {target} ({col_list}) values ({placeholders})"
        with self._conn.cursor() as cur:
            cur.executemany(sql, [tuple(_to_db_bind(value) for value in row) for row in rows])
        self._conn.commit()

    def query(self, sql: str, binds: Sequence[Any] = ()) -> list[dict[str, Any]]:
        # The harness stores golden SQL with `?` placeholders (m-sql); pymysql uses
        # `%s`. Translate positional placeholders for execution, then adapt already-managed
        # declared values and temporal-infinity markers to MariaDB forms. No raw string
        # carrier is interpreted here.
        with self._conn.cursor() as cur:
            if binds:
                cur.execute(
                    _to_pymysql(sql, escape_percent=True),
                    tuple(_to_db_bind(value) for value in binds),
                )
            else:
                cur.execute(_to_pymysql(sql))
            return _decode_rows(cur.description, cur.fetchall())

    def execute(self, sql: str, binds: Sequence[Any] = ()) -> int:
        with self._conn.cursor() as cur:
            if binds:
                affected = cur.execute(
                    _to_pymysql(sql, escape_percent=True),
                    _statement_binds(sql, binds),
                )
            else:
                affected = cur.execute(_to_pymysql(sql))
            self._conn.commit()
            return affected

    def native_error_code(self, exc: Exception) -> int | None:
        """The vendor errno pymysql packs into args[0] (else None)."""
        if exc.args and isinstance(exc.args[0], int):
            return exc.args[0]
        return None

    def classify_error(self, exc: Exception) -> str:
        """Map a raised pymysql error to a neutral m-db-error category (see errors.py)."""
        code = self.native_error_code(exc)
        return errors.UNKNOWN if code is None else _CODES.get(code, errors.UNKNOWN)

    @contextmanager
    def open_session(self) -> Iterator[_MariaTxSession]:
        """A second connection in MANUAL-commit mode, for lock contention.

        Mirrors the Postgres session. MariaDB's default
        ``innodb_lock_wait_timeout`` is 50s -- far too slow for the suite -- so it
        is lowered to 1s on open; a blocked lock then raises errno 1205 quickly.
        InnoDB detects deadlocks immediately (no timeout knob needed).
        """
        params = {**self._connect_params, "autocommit": False}
        conn = pymysql.connect(database=self._dbname, **params)
        try:
            with conn.cursor() as cur:
                cur.execute("set innodb_lock_wait_timeout = 1")
            conn.commit()
        except BaseException:
            conn.close()
            raise
        session = _MariaTxSession(conn)
        try:
            yield session
        finally:
            session.close()

    @contextmanager
    def open_peer(self) -> Iterator[Node]:
        """Yield a second, independent connection to the SAME MariaDB database.

        Cross-process coherence: node B reconnects with the provider's
        own connection parameters and ``USE``\\ s the working database, so it shares
        node A's data while holding its own session. MariaDB ``execute`` COMMITs,
        so a write on node A is visible to a read on node B — the observable half
        of cross-process coherence. Provisioning stays on node A; only the peer's
        read/write surface is used.

        The peer connects with ``autocommit=True`` so each read is its OWN
        transaction with a FRESH snapshot. Under MariaDB/InnoDB's default
        ``REPEATABLE READ`` a long-lived transaction would pin the snapshot taken
        at node B's first read and never see node A's later commit — which is the
        very thing a coherence re-fetch must observe. Autocommit reads model the
        app server's behavior exactly: a re-fetch after invalidation is a new
        query, not a read inside the stale snapshot. (Postgres' provider is already
        autocommit, so this aligns the two dialects.)
        """
        peer_params = {**self._connect_params, "autocommit": True}
        connection = pymysql.connect(database=self._dbname, **peer_params)
        peer = MariaDbProvider(connection, self._dbname, self._connect_params)
        try:
            yield peer
        finally:
            peer.close()

    def close(self) -> None:
        if self._conn is not None and self._conn.open:
            self._conn.close()


class _MariaTxSession:
    """A manual-commit MariaDB connection for two-node lock-contention cases."""

    dialect = "mariadb"

    def __init__(self, conn: pymysql.connections.Connection) -> None:
        self._conn = conn

    def execute(self, sql: str, binds: Sequence[Any] = ()) -> int:
        """Run DML on the HELD transaction; return the affected-row count.

        The count is what makes an optimistic-lock conflict observable inside a
        buffered unit of work (m-opt-lock): a stale-version gate matches 0 rows.
        The two-node lock-contention callers ignore the return.
        """
        with self._conn.cursor() as cur:
            if binds:
                cur.execute(
                    _to_pymysql(sql, escape_percent=True),
                    _statement_binds(sql, binds),
                )
            else:
                cur.execute(_to_pymysql(sql))
            return cur.rowcount

    def query(self, sql: str, binds: Sequence[Any] = ()) -> list[dict[str, Any]]:
        """Fetch rows INSIDE the held transaction (concurrency-success `expectRows`).

        Mirrors the provider's ``query`` but runs on the HELD session connection so a
        locking SELECT (``lock in share mode``) both takes the shared lock and returns
        its rows inside the open unit of work.
        """
        with self._conn.cursor() as cur:
            if binds:
                cur.execute(
                    _to_pymysql(sql, escape_percent=True),
                    tuple(_to_db_bind(value) for value in binds),
                )
            else:
                cur.execute(_to_pymysql(sql))
            return _decode_rows(cur.description, cur.fetchall())

    def commit(self) -> None:
        self._conn.commit()

    def rollback(self) -> None:
        self._conn.rollback()

    def close(self) -> None:
        if self._conn is not None and self._conn.open:
            self._conn.close()


def _to_pymysql(sql: str, *, escape_percent: bool = False) -> str:
    """Translate `?` positional placeholders to pymysql's `%s`.

    pymysql treats a literal `%` (e.g. a `like '%a%'` pattern in a naive
    referenceSql) as a format token only when an args tuple is supplied. Escape
    literal percent signs on that path, then expand the placeholders. Bindless
    SQL must keep `%` unchanged because pymysql will not format it back.
    """
    escaped = sql.replace("%", "%%") if escape_percent else sql
    return escaped.replace("?", "%s")


@contextmanager
def mariadb_provider() -> Iterator[MariaDbProvider]:
    """Boot a pinned MariaDB container and yield a provider bound to it."""
    container = MySqlContainer(
        MARIADB_IMAGE,
        # MariaDB prints a single "ready for connections" line, unlike MySQL's
        # double line the default check string expects.
        wait_strategy_check_string=r".*ready for connections.*",
    )
    container.start()
    provider: MariaDbProvider | None = None
    connection: pymysql.connections.Connection | None = None
    try:
        connect_params = _connect_params(container)
        connection = _connect_with_retry(connect_params, container.dbname)
        provider = MariaDbProvider(connection, container.dbname, connect_params)
        yield provider
    finally:
        if provider is not None:
            provider.close()
        elif connection is not None and connection.open:
            connection.close()
        container.stop()


def _connect_params(container: MySqlContainer) -> dict[str, Any]:
    """The pymysql connection parameters for the booted container (sans database).

    Shared by the provider's own connection and the peer connection, so a
    coherence case's node B reaches the same server with the same credentials.
    """
    return {
        "host": container.get_container_host_ip(),
        "port": int(container.get_exposed_port(container.port)),
        "user": container.username,
        "password": container.password,
        "autocommit": False,
    }


def _connect_with_retry(
    connect_params: dict[str, Any],
    dbname: str,
    attempts: int = 30,
    delay: float = 1.0,
) -> pymysql.connections.Connection:
    """Open a pymysql connection, retrying transient startup races.

    MariaDB's entrypoint logs ``ready for connections`` once during bootstrap and
    then restarts the server before the final ready state, so the first TCP
    handshake can be dropped (``Lost connection during query``). Retry until the
    server is genuinely accepting connections.
    """
    last_error: Exception | None = None
    for _ in range(attempts):
        try:
            return pymysql.connect(database=dbname, **connect_params)
        except pymysql.err.OperationalError as exc:  # noqa: PERF203
            last_error = exc
            time.sleep(delay)
    raise RuntimeError(f"could not connect to MariaDB after {attempts} attempts") from last_error


class _MariaDbFactory:
    """Callable + context-manager adapter so the registry can ``with factory()``."""

    def __call__(self) -> _MariaDbFactory:
        self._cm = mariadb_provider()
        return self

    def __enter__(self) -> MariaDbProvider:
        return self._cm.__enter__()

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> bool | None:
        return self._cm.__exit__(exc_type, exc, tb)


register("mariadb", _MariaDbFactory())

File: languages/typescript/packages/dialect/src/mariadb.ts
/**
 * The M11 MariaDB dialect — the **second** conforming {@link Dialect}.
 *
 * MariaDB exercises exactly the seam points Postgres does not, proving the
 * abstraction earns its keep at the SQL/value-rule level (the corpus divergences
 * `0006`, `0323`, `1001`, `1002`, `1005`, `0720`–`0727`):
 *
 *  - **identifier quoting** — a backtick, not a double-quote (`` `order` ``), with
 *    embedded backticks doubled and its own reserved-word set;
 *  - **NULL ordering** — MariaDB has no `NULLS LAST` syntax, so an ascending
 *    nullable key forces NULLs last with a leading `is null,` term (`m11:153-167`);
 *  - **read-lock** — MariaDB has no `for share` (MDEV-17514); its shared row lock
 *    is the unaliased ` lock in share mode`;
 *  - **neutral→column type** — `tinyint(1)` / `datetime(6)` / `double` / `longblob`
 *    / `char(36)` … (the M0 table, its MariaDB column);
 *  - **infinity** — MariaDB's `DATETIME` has no native `'infinity'`, so the open
 *    temporal upper bound binds a documented MAX-SENTINEL datetime (`m11:189-202`);
 *  - **error classification** — MariaDB keys on the vendor errno (`1062`/`1213`/
 *    `1205`), not a SQLSTATE string;
 *  - **placeholders** — the MariaDB driver takes native `?`, so the adapter's
 *    `?`→positional rewrite is the identity.
 *
 * Like `postgresDialect`, this module imports no driver: it exposes the *rules* a
 * driver-bound adapter (`@parallax/db-mariadb`, a later phase) applies. Value
 * parsers reuse the neutral `@parallax/core` scalar helpers; only `timestamp`
 * (max-sentinel ↔ infinity) diverges from the Postgres wire form.
 */
import {
  bytesFromHex,
  INFINITY,
  type Infinity as InfinitySentinel,
  ParallaxDecimal,
  Temporal,
  timestampFromRaw,
} from "@parallax/core";
import type { Dialect } from "./dialect.js";
import type { ErrorCategory } from "./errors.js";

/** The dialect identifier this seam answers for (keys `goldenSql`/`expectedNativeCode`). */
export const MARIADB_DIALECT = "mariadb" as const;

/**
 * The MariaDB max-sentinel for the open temporal upper bound: the largest
 * representable `DATETIME(6)`. MariaDB's `DATETIME` has no native `'infinity'`, so
 * the seam substitutes this documented sentinel for the neutral `infinity` on the
 * way in (binds) and detects it on the way out (parsers), so a fixture authored
 * once against native-infinity Postgres compares identically here (`m11:189-202`).
 */
export const MARIADB_INFINITY_SENTINEL = "9999-12-31 23:59:59.999999" as const;

/**
 * MariaDB takes native `?` positional placeholders, so the canonical M3 `?` SQL
 * needs no rewrite at the driver boundary — this is the identity (contrast the
 * Postgres `?`→`$n` translation).
 */
function toPositionalPlaceholders(sql: string): string {
  return sql;
}

/**
 * **Apply** MariaDB's in-transaction shared read lock (M8 automatic read-lock
 * correctness, owned by the M11 seam). MariaDB has no `for share` (MDEV-17514); its
 * shared row lock is the unaliased ` lock in share mode`, appended after every
 * other clause. Mirrors the Postgres decision structure exactly — a projection /
 * aggregation read (`projection: true`) and any non-`locking` read are returned
 * unchanged (no base row to lock) — only the suffix text diverges.
 */
function applyReadLock(
  sql: string,
  options: { readonly locking: boolean; readonly projection: boolean },
): string {
  // MariaDB's shared lock is UNALIASED (unlike Postgres's `for share of t0`).
  if (!options.locking || options.projection) {
    return sql;
  }
  return `${sql} lock in share mode`;
}

/**
 * One ORDER BY term with MariaDB's NULL placement (the M4/M11 catalog table,
 * `m11:153-167`). MariaDB has no `NULLS LAST` syntax and sorts NULLs FIRST for
 * `asc` / LAST for `desc` by default, so:
 *
 *  - `asc` forces NULLs last with a leading `<col> is null,` term (`is null` is `0`
 *    for present values and `1` for NULLs, so ascending on it trails the NULLs)
 *    followed by the plain ascending sort — the leading term becomes part of the
 *    ORDER BY list the compiler joins with `, `;
 *  - `desc` is emitted **bare** (its native default already sorts NULLs last, so no
 *    extra term is needed to match the neutral "NULLs sort last" rule).
 *
 * Postgres expresses the same intent with a bare `asc` / an explicit `desc nulls
 * last` (its `postgresDialect` sibling).
 */
function orderByTerm(qualifiedColumn: string, direction: "asc" | "desc"): string {
  if (direction === "desc") {
    // Bare `desc` — MariaDB sorts NULLs LAST for descending by default, matching the
    // M4 canonical rule. Do NOT add `nulls last` (MariaDB has no such syntax).
    return `${qualifiedColumn} desc`;
  }
  // A leading `is null,` term forces NULLs LAST for ascending (MariaDB's default is
  // NULLs FIRST). This is spec-mandated (`m11:153-167`); do NOT drop it.
  return `${qualifiedColumn} is null, ${qualifiedColumn} asc`;
}

/**
 * Apply the MariaDB row-limit clause — append ` limit ?` (the bind carries the
 * count). Byte-identical to Postgres today; modeled as a *wrappable* hook so a
 * future dialect that rewrites the query shape can override rather than append.
 */
function rowLimit(sql: string): string {
  return `${sql} limit ?`;
}

/**
 * Lower a `bytes` column to MariaDB's hex-text projection — the argument-less
 * `hex(<col>) <out>` — carrying NO bind (contrast Postgres's `encode(<col>, ?)`
 * with a `'hex'` format bind). MariaDB's `hex(...)` takes no format argument, so
 * the projection is bind-free (`1005`).
 */
function bytesProjection(
  qualifiedColumn: string,
  outputName: string,
): { readonly sql: string; readonly binds: readonly unknown[] } {
  return { sql: `hex(${qualifiedColumn}) ${outputName}`, binds: [] };
}

// --- neutral-type → MariaDB column type (the M0 table) ----------------------

/**
 * M0 neutral base type → MariaDB column type (non-parametric types). The
 * divergences from Postgres that matter: `boolean`→`tinyint(1)` (no native
 * boolean), `timestamp`→`datetime(6)` (MariaDB `TIMESTAMP` is 2038-limited +
 * auto-updates, so milestones use `DATETIME` with µs precision — and it has no
 * native infinity), `float64`→`double`, `bytes`→`longblob`, `uuid`→`char(36)`.
 */
const MARIADB_BASE_TYPES: Readonly<Record<string, string>> = {
  boolean: "tinyint(1)",
  int32: "int",
  int64: "bigint",
  float32: "float",
  float64: "double",
  bytes: "longblob",
  date: "date",
  time: "time",
  timestamp: "datetime(6)",
  uuid: "char(36)",
  json: "json",
};

/** Matches a `decimal(p,s)` neutral type token and captures precision/scale. */
const DECIMAL_TYPE = /^decimal\((\d+),(\d+)\)$/;

/**
 * Map an M0 neutral type to its MariaDB column type. `decimal(p,s)` stays
 * `decimal(p,s)` (MariaDB spells it `decimal`, not `numeric`), a bounded `string`
 * → `varchar(n)`, an unbounded `string` → `text`; everything else comes from the
 * base table.
 */
function mariadbColumnType(neutralType: string, maxLength?: number): string {
  const decimal = DECIMAL_TYPE.exec(neutralType);
  if (decimal) {
    return `decimal(${decimal[1]},${decimal[2]})`;
  }
  if (neutralType === "string") {
    return maxLength ? `varchar(${maxLength})` : "text";
  }
  const base = MARIADB_BASE_TYPES[neutralType];
  if (base === undefined) {
    throw new Error(`no MariaDB mapping for neutral type '${neutralType}'`);
  }
  return base;
}

// --- identifier quoting ------------------------------------------------------

/** A lexically-simple identifier needs no quoting unless it is reserved. */
const SIMPLE_IDENTIFIER = /^[a-z_][a-z0-9_]*$/;

/**
 * Reserved words that, although lexically simple, MUST be quoted as an identifier.
 * MariaDB carries its OWN reserved set (a database's keyword list differs from
 * Postgres's); the curated set below covers the identifiers the corpus models
 * emit (e.g. `order`) — enough to keep generated SQL byte-identical to the
 * hand-authored goldens. A non-simple name (uppercase / special) is caught by the
 * regex regardless.
 *
 * `position` is a MariaDB-only addition (`POSITION()` is a reserved SQL function
 * name here but not on Postgres, and the corpus's `Position` table would otherwise
 * emit unquoted on MariaDB — both in this DDL derivation and in the M3 compiler's
 * `from` clause, which now also routes through `quoteIdentifier`). Postgres's
 * reserved set intentionally omits it (`position t0` stays unquoted there,
 * byte-identical to the hand-authored golden).
 */
const RESERVED_WORDS: ReadonlySet<string> = new Set([
  "all",
  "and",
  "as",
  "asc",
  "between",
  "by",
  "case",
  "check",
  "column",
  "constraint",
  "create",
  "default",
  "delete",
  "desc",
  "distinct",
  "drop",
  "else",
  "end",
  "exists",
  "foreign",
  "from",
  "group",
  "having",
  "in",
  "index",
  "insert",
  "into",
  "is",
  "join",
  "key",
  "like",
  "limit",
  "not",
  "null",
  "on",
  "or",
  "order",
  "position",
  "primary",
  "references",
  "select",
  "set",
  "table",
  "then",
  "to",
  "union",
  "unique",
  "update",
  "user",
  "using",
  "values",
  "when",
  "where",
]);

/**
 * Quote a MariaDB identifier when it is reserved or otherwise non-simple. A simple
 * lowercase non-reserved identifier is returned unquoted (so generated SQL is
 * byte-identical to the hand-authored goldens); a reserved word (e.g. `order`) or
 * a name with uppercase / special characters is **backtick**-quoted, with any
 * embedded backtick doubled — the one genuine cross-dialect divergence in the
 * quote CHARACTER (Postgres double-quotes; MariaDB backticks).
 */
function quoteIdentifier(name: string): string {
  if (SIMPLE_IDENTIFIER.test(name) && !RESERVED_WORDS.has(name)) {
    return name;
  }
  return `\`${name.replace(/`/g, "``")}\``;
}

// --- raw-string type coercion at the adapter boundary -----------------------

/** Materialize a raw `int8` string into a native `bigint` (µ-precision-safe). */
function int8FromRaw(raw: string): bigint {
  return BigInt(raw.trim());
}

/** Materialize a raw `numeric` string into an exact {@link ParallaxDecimal}. */
function numericFromRaw(raw: string): ParallaxDecimal {
  return ParallaxDecimal.from(raw.trim());
}

/**
 * Materialize a raw MariaDB `datetime(6)` string into a `Temporal.Instant` at
 * microsecond precision. The MAX-SENTINEL (`9999-12-31 23:59:59.999999`) is mapped
 * back to the `infinity` sentinel — the open temporal upper bound has no instant —
 * so a current-row open bound compares to the fixture's native-infinity Postgres
 * value. A MariaDB `DATETIME` carries no offset; `timestampFromRaw` treats an
 * offset-less value as UTC (every instant in the suite is UTC).
 */
function timestampFromDb(raw: string): Temporal.Instant | InfinitySentinel {
  const text = raw.trim();
  if (text === MARIADB_INFINITY_SENTINEL) {
    return INFINITY;
  }
  return timestampFromRaw(text);
}

/**
 * Materialize a raw `longblob` hex rendering into a `Uint8Array`. MariaDB's
 * `hex(...)` yields plain hex text (no `\x` prefix); {@link bytesFromHex} accepts
 * both the prefixed and bare forms, so the same parser serves both dialects.
 */
function bytesFromDb(raw: string): Uint8Array {
  return bytesFromHex(raw.trim());
}

/** Materialize a raw `date` string (`2024-03-01`) into a `Temporal.PlainDate`. */
function dateFromDb(raw: string): Temporal.PlainDate {
  return Temporal.PlainDate.from(raw.trim());
}

/** Materialize a raw `time` string (`12:34:56`) into a `Temporal.PlainTime`. */
function timeFromDb(raw: string): Temporal.PlainTime {
  return Temporal.PlainTime.from(raw.trim());
}

/** Materialize a raw `char(36)` uuid into a canonical string. */
function uuidFromDb(raw: string): string {
  return raw.trim();
}

// --- error classification (MariaDB vendor errno) ----------------------------

/**
 * MariaDB keys error classification on the vendor **errno** (an int), not the
 * SQLSTATE string Postgres uses: `1062` (ER_DUP_ENTRY), `1213` (ER_LOCK_DEADLOCK),
 * `1205` (ER_LOCK_WAIT_TIMEOUT). This is exactly why the code *source* is a dialect
 * decision — SQLSTATE `40001` is a serialization failure on Postgres but the
 * deadlock state on MariaDB (`errors.py` records the same split).
 */
const MARIADB_ERROR_CODES: Readonly<Record<number, ErrorCategory>> = {
  1062: "uniqueViolation",
  1213: "deadlock",
  1205: "lockWaitTimeout",
};

/**
 * Classify a native MariaDB error code (the vendor errno the driver surfaces) to a
 * neutral M11 category. Coerces the code to an integer errno and looks it up;
 * returns `unknown` for an unrecognized / missing / non-numeric code so an
 * unclassified error is never silently treated as retriable.
 */
function classifyErrorCode(code: string | number | null | undefined): ErrorCategory {
  if (code === null || code === undefined) {
    return "unknown";
  }
  const errno = Number(code);
  if (!Number.isInteger(errno)) {
    return "unknown";
  }
  return MARIADB_ERROR_CODES[errno] ?? "unknown";
}

// --- the reified MariaDB dialect ---------------------------------------------

/**
 * The concrete MariaDB {@link Dialect} — the second conforming implementation of
 * the layer-1 contract. Every method is MariaDB's answer to the same catalog
 * question `postgresDialect` answers for Postgres; the two diverge only where the
 * corpus witnesses a genuine dialect difference. `infinityBind` returns the
 * max-sentinel datetime MariaDB binds for the open temporal upper bound.
 */
export const mariadbDialect: Dialect = {
  id: MARIADB_DIALECT,
  quoteIdentifier,
  orderByTerm,
  rowLimit,
  bytesProjection,
  applyReadLock,
  columnType: mariadbColumnType,
  toPositionalPlaceholders,
  parsers: {
    int8: int8FromRaw,
    numeric: numericFromRaw,
    timestamp: timestampFromDb,
    bytes: bytesFromDb,
    date: dateFromDb,
    time: timeFromDb,
    uuid: uuidFromDb,
  },
  infinityBind: () => MARIADB_INFINITY_SENTINEL,
  classifyErrorCode,
  isRetriable: (category) => category === "deadlock",
  violatesUniqueIndex: (category) => category === "uniqueViolation",
  isTimedOut: (category) => category === "lockWaitTimeout",
};
